pkg/cloudprovider/cloudprovider.go

```diff
@@ -316,6 +316,7 @@ func (c *CloudProvider) instanceToNodeClaim(i *instance.Instance, instanceType *
 		nodeClaim.Status.Allocatable = functional.FilterMap(instanceType.Allocatable(), resourceFilter)
 	}
 	labels[v1.LabelTopologyZone] = i.Zone
+	labels[v1beta1.LabelTopologyZoneID] = c.subnetProvider.ZoneInfo()[i.Zone]
```

Comment:
I didn't think this raced because I thought pointer writes were atomic, but it doesn't look like that's guaranteed by the spec. As for why it's not on Instance, @rschalo can correct me if I'm wrong but I think the main reason is because it's not available as part of the DescribeInstances output. We'd have to populate it from either the EC2NodeClass or the subnet provider inside NewInstance. IMO this is the cleaner approach.

Reply:
Correct that the ID is not returned as part of the DescribesInstances call. Before we return the instance from the instance provider, we could add the ID there. Rather than reset it every time we describe subnets, would it work better to build out all of the zones for the account? That way, even if the subnets change, the account specific mapping remains.

Additionally, we could make an EC2 feature request for this to be returned as part of this request.

Reply:
Came up with a slightly different solution, we're relying completely on resolution from the NodeClass. In the create path (where we really need this label) we've already validated that the nodeclass exists and can continue to leverage it. In the List and Get are only used for garbage collection and node termination respectively, in which case it doesn't matter if we can't populate this label because we can't resolve the nodeclass. IMO this is a good solution which eliminates the need for tracking the zone-id mapping in the subnet provider at all.